src/qibo/models.py

```diff
@@ -198,70 +199,178 @@ def loss(params):
 
 
 class QAOA(object):
-    """ This class implements the QAOA algorithm."""
-    import numpy as np
-    from qibo import hamiltonians
-        
-    def __init__(self, nqubits, hamiltonian, mixer=None, initial_state=None):
-        """
-        Initializes an instance of the QAOA.
-        
-        Args:
-            nqubits (int): number of qubits.
-            hamiltonian (qibo.hamiltonians.Hamiltonian): problem Hamiltonian whose ground state is sought.
-            mixer (qibo.hamiltonians.Hamiltonian): mixer Hamiltonian. If None, it is taken as -sum(sigma_x).
-            initial state: initial state of the algorithm. If None, it is taken as the uniform
-                           superposition over computational-basis states.
-        """
+    """ Quantum Approximate Optimization Algorithm (QAOA) model.
+
+    The QAOA is introduced in `arXiv:1411.4028 <https://arxiv.org/abs/1411.4028>`_.
+
+    Args:
+        hamiltonian (:class:`qibo.base.hamiltonians.Hamiltonian`): problem Hamiltonian
+            whose ground state is sought.
+        mixer (:class:`qibo.base.hamiltonians.Hamiltonian`): mixer Hamiltonian.
+            If ``None``, :class:`qibo.hamiltonians.X` is used.
+        solver (str): solver used to apply the exponential operators.
+        callbacks (list): List of callbacks to calculate during evolution.
+        accelerators (dict): Dictionary of devices to use for distributed
+            execution. See :class:`qibo.tensorflow.distcircuit.TensorflowDistributedCircuit`
+            for more details. This option is available only when ``hamiltonian``
+            is a :class:`qibo.base.hamiltonians.TrotterHamiltonian`.
+        memory_device (str): Name of device where the full state will be saved.
+            Relevant only for distributed execution (when ``accelerators`` is
+            given).
+
+    Example:
+        ::
+
+            import numpy as np
+            from qibo import models, hamiltonians
+            # create XXZ Hamiltonian for four qubits
+            hamiltonian = hamiltonians.XXZ(4)
+            # create QAOA model for this Hamiltonian
+            qaoa = models.QAOA(hamiltonian)
+            # optimize using random initial variational parameters
+            # and default options and initial state
+            initial_parameters = 0.01 * np.random.random(4)
+            best_energy, final_parameters = qaoa.minimize(initial_parameters, method="BFGS")
+    """
+    from qibo import hamiltonians, optimizers
+    from qibo.config import K, DTYPES, log
+    from qibo.callbacks import Norm
+
+    def __init__(self, hamiltonian, mixer=None, solver="exp", callbacks=[],
+                 accelerators=None, memory_device="/CPU:0"):
+        # list of QAOA variational parameters (angles)
+        self.params = None
         # problem hamiltonian
         self.hamiltonian = hamiltonian
```

Comment:
Shall we check that hamiltonian is really a hamiltonian object?

Reply:
I added a check for this.